### PR TITLE
consumerRestriction: add method support

### PR DESCRIPTION
**Pull request by hanxiantao**
fixes: https://github.com/mosn/htnn/issues/253

### CODE REVIEW

File: plugins/plugins/consumer_restriction/config.go [hunk 1]
```diff
@@ -50,9 +55,16 @@ func (conf *config) Init(cb api.ConfigCallbackHandler) error {
 		conf.allow = true
 	}
 
-	conf.consumers = make(map[string]struct{}, len(rules.Rules))
+	conf.consumers = make(map[string]*Rule, len(rules.Rules))
 	for _, r := range rules.Rules {
-		conf.consumers[r.Name] = struct{}{}
+		methods := make(map[string]bool)
+		for _, method := range r.GetMethods() {
+			methods[method] = true
```
Comment (spacewander) [inline plugins/plugins/consumer_restriction/config.go:62]: Maybe we can add a validation to make sure the method is uppercase?
https://github.com/bufbuild/protoc-gen-validate?tab=readme-ov-file#repeated

File: site/content/zh-hans/docs/developer-guide/plugin_development.md [hunk 2]
```diff
@@ -15,7 +15,7 @@ HTNN 的插件分成两种：Go 插件和 Native 插件。Native 插件在运行
 5. 在 `./plugins/plugins/` 下创建一个目录，文件名必须和步骤一里面的名字一致。完成插件编写。不要忘记编写测试。如果您的插件简单，您可以仅编写集成测试。您可以以 `./plugins/plugins/demo` 为例。插件开发中涉及的 API 的文档位于这些 API 的注释当中。
 6. 在 `site/content/$your_language/docs/reference/plugins/$your_plugin.md` 中添加插件文档。您可以选择使用简体中文或英文编写文档，这取决于哪种是您擅长的语言。我们有 [工具](https://github.com/mosn/htnn/tree/main/site#cmdtranslator) 可以将其翻译成其他语言。
 7. 在 `./plugins/plugins.go` 中引入你的插件。前往目录 `./plugins`，运行 `make build-so`。插件将被编译入当前目录下的 `libgolang.so`。
-8. 在 `./plugins/tests/integration/` 中添加集成测试。有关如何运行集成测试，请阅读[插件集成测试框架](../plugin_integration_test_framework)。
+8. 在 `./plugins/tests/integration/` 中添加集成测试。有关如何运行集成测试，请阅读[插件集成测试框架](./plugin_integration_test_framework)。
```
Comment (spacewander) [inline site/content/zh-hans/docs/developer-guide/plugin_development.md:18]: This is a bit tricky. The doc generated by the site builder is different from what it looks like on GitHub. You could take a look at the output result by building the website: https://github.com/mosn/htnn/blob/de1d13b57a48881edf83c1a03d62a2d38a1b821f/.github/workflows/website.yml#L39.

File: site/content/en/docs/developer-guide/plugin_integration_test_framework.md [hunk 3]
```diff
@@ -7,7 +7,8 @@ title: Plugin Integration Test Framework
 Assumed you are at the `./plugins`:
 
 1. Run `make build-test-so` to build the Go plugins.
-2. Run `go test -v ./tests/integration -run TestPluginXX` to run the selected tests.
+2. Place the built `libgolang.so` into the `./tests/integration` directory.
```
Comment (spacewander) [inline site/content/en/docs/developer-guide/plugin_integration_test_framework.md:10]: Interesting. The `make build-test-so` should already place `libgolang.so` into the `./tests/integration` directory:
https://github.com/mosn/htnn/blob/de1d13b57a48881edf83c1a03d62a2d38a1b821f/plugins/Makefile#L33

Why doesn't it work?

File: plugins/plugins/consumer_restriction/filter.go [hunk 4]
```diff
@@ -43,10 +43,18 @@ func (f *filter) DecodeHeaders(headers api.RequestHeaderMap, endStream bool) api
...
-	_, ok := f.config.consumers[consumer.Name()]
-	if ok != f.config.allow {
-		api.LogInfof("consumerRestriction: consumer %s not allowed", consumer.Name())
-		// don't leak consumer name to the caller
+	consumerName := consumer.Name()
+	rule, consumerExists := f.config.consumers[consumerName]
+	reqMethod := headers.Method()
+	methodAllowed := rule != nil && (len(rule.Methods) == 0 || rule.Methods[reqMethod])
+
+	if f.config.allow {
```
Comment (spacewander) [inline plugins/plugins/consumer_restriction/filter.go:51]: Look like we can simplify the comparison as `methodMatched != f.config.allow`.
If `!methodAllowed` is true then `!consumerExists` must be true.

File: site/content/zh-hans/docs/developer-guide/plugin_development.md (hunk 2 above)
Comment (spacewander) [reply site/content/zh-hans/docs/developer-guide/plugin_development.md:18]: Interesting. Both the version in this PR and the version in the main branch can pass the link check. (We have a link checker in the CI, you can learn more from https://github.com/mosn/htnn/pull/600)

Look like an absolute link is generated:
```
<li>在 <code>./plugins/tests/integration/</code> 中添加集成测试。有关如何运行集成测试，请阅读<a href="/zh-hans/docs/developer-guide/plugin_integration_test_framework/">插件集成测试框架</a>。</li>
```

Could you put this change in another PR?

File: plugins/plugins/consumer_restriction/config.go (hunk 1 above)
Comment (hanxiantao) [reply plugins/plugins/consumer_restriction/config.go:62]: > Maybe we can add a validation to make sure the method is uppercase? https://github.com/bufbuild/protoc-gen-validate?tab=readme-ov-file#repeated

add a validation in types/plugins/consumer_restriction/config.proto

File: site/content/zh-hans/docs/developer-guide/plugin_development.md (hunk 2 above)
Comment (hanxiantao) [reply site/content/zh-hans/docs/developer-guide/plugin_development.md]: > Interesting. Both the version in this PR and the version in the main branch can pass the link check. (We have a link checker in the CI, you can learn more from #600)
> 
> Look like an absolute link is generated:
> 
> ```
> <li>在 <code>./plugins/tests/integration/</code> 中添加集成测试。有关如何运行集成测试，请阅读<a href="/zh-hans/docs/developer-guide/plugin_integration_test_framework/">插件集成测试框架</a>。</li>
> ```
> 
> Could you put this change in another PR?

Thank you, I will go and check out this PR.

File: site/content/en/docs/reference/plugins/consumer_restriction.md [hunk 5]
```diff
@@ -33,7 +33,7 @@ Only one of `allow` or `deny` can be configured.
 | Name | Type   | Required | Validation   | Description          |
 |------|--------|----------|--------------|----------------------|
 | name | string | True     | min_len: 1   | Name of the Consumer |
-
+| methods | string[] | False   |  | List of HTTP methods allowed/prohibited for Consumer |
```
Comment (spacewander) [inline site/content/en/docs/reference/plugins/consumer_restriction.md:36]: Would you add the pattern to "Validation" before I can merge it? Thanks!

Comment (hanxiantao) [reply site/content/en/docs/reference/plugins/consumer_restriction.md]: > Would you add the pattern to "Validation" before I can merge it? Thanks!

The validation is annotated with 'must be uppercase'

**Review by spacewander — APPROVED** [review]

**Comment by spacewander** [comment]
Merged. Thanks!

**Pull request merged**